### PR TITLE
Switch from extra-data to normal package (v2)

### LINKED ISSUE #42: Flatpak downloads every nvidia driver from flathub at every update
Issue moved from here: https://github.com/flatpak/flatpak/issues/3750

## Linux distribution and version

Ubuntu 20.04

## Flatpak version

1.6.3

## Description of the problem

Every time a new nvidia driver gets released I need to update the flatpaks and this already slow process(with SSD and very fast internet) becomes slower because it redownloads every driver again. This happened three times so far. Now it downloaded every driver from 430.xy to 450.xy.

## Steps to reproduce

Run `flatpak update` when there is a new nvidia driver available.

### LINKED ISSUE #74: Connection slowly dying
I don't undesrtand why only this package cannot update, here's the process (it's slow enough to be able to copy):
```
flatpak update
Looking for updates…


        ID                                                  Branch        Op        Remote         Download
 1.     org.freedesktop.Platform.GL.nvidia-470-63-01        1.4           i         flathub        < 223,4 MB

Proceed with these changes to the user installation? [Y/n]: 
```
Y
```
        ID                                                  Branch        Op        Remote         Download
 1. [\] org.freedesktop.Platform.GL.nvidia-470-63-01        1.4           i         flathub        1,0 kB / 223,4 MB

Installing… █▌                     8%  197 bytes/s

```
it starts at 200b/s and goes down, and it finally ends with:
```
        ID                                                  Branch        Op        Remote         Download
 1. [✗] org.freedesktop.Platform.GL.nvidia-470-63-01        1.4           i         flathub        1,0 kB / 223,4 MB

Warning: While downloading https://us.download.nvidia.com/XFree86/Linux-x86_64/470.63.01/NVIDIA-Linux-x86_64-470.63.01-no-compat32.run: Could not connect: No route to host
Installation complete.

```

"Installation complete"? Not really :)

Current flatpak source is https://dl.flathub.org/repo
I'm using optical fiber (300MBps, solid 30MBps) in Europe. Maybe this Nvidia server route is not for Europe?

Thanks for your lights on this.

### LINKED ISSUE #156: Avoid downloading the same driver twice on every update
Hi, currently everytime there is a nvidia update, it does this:
```
 1. [✓] org.freedesktop.Platform.GL.nvidia-525-85-05                1.4                 u              flathub                 415,1 MB / 415,9 MB
 2. [✓] org.freedesktop.Platform.GL32.nvidia-525-85-05              1.4                 u              flathub                 415,1 MB / 415,9 MB
```

It downloads this 415,1 MB file from nvidia, for 64 and 32 bit.
However afaik these are the same file, quote from update-data.sh:
```
# Nvidia dropped 32bit support after 390 series but 64bit contains
# 32bit compat libs
```

This is very frustrating especially on slow connections since it takes ages... twice.

An idea would be to probably combine these to packages into one.
So that we only have `org.freedesktop.Platform.GL.nvidia...` but not the `GL32` one anymore.
This would solve two problems at once:
1.) the double download of the driver files
2.) the fact that `flatpak remove --unused` currently does not remove the `GL32` drivers extension, see:
https://github.com/flatpak/flatpak/issues/2718

If not possible, is there at least some way to download the driver only once and reuse the file?

### LINKED ISSUE #203: Get permission from Nvidia to redistribute drivers directly?
At the moment, this driver uses Flatpak's "extra data" mechanism, which stops working if Nvidia stops hosting a particular driver (for example https://github.com/flatpak/flatpak/issues/4810), and can be unnecessarily slow if the user's connection to Nvidia's servers is worse than their connection to Flathub (or a mirror of Flathub).

A better scenario for these drivers would be if the maintainers of this driver bundle, or even Nvidia themselves, could upload the driver to Flathub as an ordinary Flatpak extension, the same way Mesa works.

A while ago I put some people from Flathub in contact with developers in Nvidia, in the hope that they could negotiate a way to get permission to redistribute the Nvidia drivers through Flathub without using "extra data". For example, Debian got similar permission 20 years ago so that they could ship the drivers in a .deb. Did anything ever come out of that?

### LINKED ISSUE #217: apply_extra script failed with org.freedesktop.Platform.GL.nvidia-545-29-06
Trying to install `org.freedesktop.Platform.GL.nvidia-545-29-06` results in an error:

```
$ flatpak install org.freedesktop.Platform.GL.nvidia-545-29-06
Looking for matches…


        ID                                                       Branch             Op            Remote             Download
 1. [✗] org.freedesktop.Platform.GL.nvidia-545-29-06             1.4                i             flathub            324.7 MB / 325.6 MB

Warning: Cannot create repo on revokefs mountpoint /var/tmp/flatpak-cache-KTV4E2/org.freedesktop.Platform.GL.nvidia-545-29-06-SUFFF2: mkdirat: Input/output error
Error: Error deploying: While trying to apply extra data: apply_extra script failed, exit status 256
error: Failed to install org.freedesktop.Platform.GL.nvidia-545-29-06: Error deploying: While trying to apply extra data: apply_extra script failed, exit status 256
```

Not sure how to debug this further.

### PR DESCRIPTION
**Pull request by guihkx**
This is my take at solving #203, which is also pretty much a follow-up to #167.

Here are some pros of accepting this PR (in my opinion):

- Less annoying driver updates ([#42](https://github.com/flathub/org.freedesktop.Platform.GL.nvidia/issues/42), [#74](https://github.com/flathub/org.freedesktop.Platform.GL.nvidia/issues/74), [#156](https://github.com/flathub/org.freedesktop.Platform.GL.nvidia/issues/156))
- Bypass of potential regional restrictions: ([#203](https://github.com/flathub/org.freedesktop.Platform.GL.nvidia/issues/203#issuecomment-1741100717), [#52](https://github.com/flathub/org.freedesktop.Platform.GL.nvidia/issues/52))
- The retirement of the `i386` build bot

And here are some potential cons (also in my opinion):

- Potential throttling of download speeds by NVIDIA servers: I built every single driver version we currently have, just to make sure they would at least pass the build step. However, after downloading about half of the drivers at very high speeds (50 MB/s), subsequent downloads were slowed down to only 100 KB/s (!). I made sure to check that this issue wasn't being caused by my ISP, too. So, I'm not sure if Flathub servers will be affected by this, but it's something to keep in mind, because it might significantly slow down the build process.
- Overall slower build time: This issue is not directly related to the issue above (although this one can be worsened by it), Instead, it has to do with the fact that by switching from extra-data, Flathub now has to download 79 GiB worth of driver files _every time_ a new driver version is added. A potential way mitigate this, would be by Flathub admins allocating ~100 GB of storage to cache all these installers, but this number will only grow with time... :(

    ```bash
    # Estimation of how much GiBs we have to download to build every single driver (aarch64 drivers included)
    $ cat data/* | uniq | awk -F ':' '{sum+=$3}END{print sum/1024/1024/1024" GiB"}'
    79.0474 GiB
    ```

That being said, please also note that my testing capabilities are limited because I only have an old NVIDIA GPU, and so the *actual testing* was solely done with version 470.223.02 of the driver.

In any case, I have wrote some instructions below for any contributors willing to build and test this locally.

Closes #42
Closes #74
Closes #156
Closes #203
Closes #217

---

### How to build/package locally

1\. Open `versions.sh` and edit the `DRIVER_VERSIONS` variable at the very end, informing the versions you want to package, separated by a whitespace. For example, to package _only_ the `545.29.06` driver:

```bash
DRIVER_VERSIONS='545.29.06'
```

2\. Save the file.
3\. Now simply invoke `make` to build for `x86_64` (and `i386`, simultaneously):

```bash
# hint: you can build for aarch64 by adding ARCH=aarch64 in front of make
make
```

4\. After building finishes, we can now create [Flatpak bundles](https://docs.flatpak.org/en/latest/single-file-bundles.html), so we can easily install the drivers later:

```bash
# 64-bit driver
flatpak build-bundle repo/ nvidia-driver-545.29.06-64bit.flatpak org.freedesktop.Platform.GL.nvidia-545-29-06 1.4 --runtime
# 32-bit driver (only available if building for x86_64)
flatpak build-bundle repo/ nvidia-driver-545.29.06-32bit.flatpak org.freedesktop.Platform.GL32.nvidia-545-29-06 1.4 --runtime
```

5\. And then you can install them with:

```bash
flatpak install --bundle nvidia-driver-545.29.06-64bit.flatpak
flatpak install --bundle nvidia-driver-545.29.06-32bit.flatpak
```

### Testing

Perhaps the easiest way to test this would be installing [Flatpak'd Steam](https://flathub.org/apps/com.valvesoftware.Steam), since it already requires the `multiarch` permission, and so it's much easier to also test the 32-bit part of the drivers.

My tests consist in downloading x86 and x86_64 binaries of `vulkaninfo` (to test Vulkan) and `glxinfo` (to test OpenGL), and then running these two tools within the `com.valvesoftware.Steam` app, and observe their output is correct.

Below are command-line only instructions on how to do that.

#### Testing Vulkan (x86/x86_64)

```bash
# RPM source: https://kojihub.stream.centos.org/koji/buildinfo?buildID=46408
curl -LO https://kojihub.stream.centos.org/kojifiles/packages/vulkan-tools/1.3.268.0/1.el9/x86_64/vulkan-tools-1.3.268.0-1.el9.x86_64.rpm
# Note: You'll need rpmextract to extract vulkaninfo
rpm2cpio vulkan-tools-1.3.268.0-1.el9.x86_64.rpm | zstd -d | cpio -i --to-stdout ./usr/bin/vulkaninfo > vulkaninfo64
# Give it execute permission
chmod +x vulkaninfo64
# Run vulkaninfo within Flatpak'd Steam, and make sure that the output matches your GPU name/model, and not llvmpipe
flatpak run --command=$PWD/vulkaninfo64 --filesystem=$PWD com.valvesoftware.Steam --summary | grep -wm1 deviceName
```

```bash
# RPM source: https://kojihub.stream.centos.org/koji/buildinfo?buildID=46408
curl -LO https://kojihub.stream.centos.org/kojifiles/packages/vulkan-tools/1.3.268.0/1.el9/i686/vulkan-tools-1.3.268.0-1.el9.i686.rpm
# Note: You'll need rpmextract to extract vulkaninfo
rpm2cpio vulkan-tools-1.3.268.0-1.el9.i686.rpm | zstd -d | cpio -i --to-stdout ./usr/bin/vulkaninfo > vulkaninfo32
# Give it execute permission
chmod +x vulkaninfo32
# Run vulkaninfo within Flatpak'd Steam, and make sure that the output matches your GPU name/model, and not llvmpipe
flatpak run --command=$PWD/vulkaninfo32 --filesystem=$PWD com.valvesoftware.Steam --summary | grep -wm1 deviceName
```

#### Testing OpenGL (x86/x86_64)

```bash
# RPM source: https://kojihub.stream.centos.org/koji/buildinfo?buildID=19862
curl -LO https://kojihub.stream.centos.org/kojifiles/packages/mesa-demos/8.4.0/5.20181118git1830dcb.el8/x86_64/glx-utils-8.4.0-5.20181118git1830dcb.el8.x86_64.rpm
# Note: You'll need rpmextract to extract glxinfo
rpm2cpio glx-utils-8.4.0-5.20181118git1830dcb.el8.x86_64.rpm | zstd -d | cpio -i --to-stdout ./usr/bin/glxinfo > glxinfo64
# Give it execute permission
chmod +x glxinfo64
# Run glxinfo within Flatpak'd Steam, and make sure that the does not say 'llvmpipe'
flatpak run --command=$PWD/glxinfo64 --filesystem=$PWD com.valvesoftware.Steam -B | grep 'OpenGL renderer string:'
```

```bash
# RPM source: https://kojihub.stream.centos.org/koji/buildinfo?buildID=19862
curl -LO https://kojihub.stream.centos.org/kojifiles/packages/mesa-demos/8.4.0/5.20181118git1830dcb.el8/i686/glx-utils-8.4.0-5.20181118git1830dcb.el8.i686.rpm
# Note: You'll need rpmextract to extract glxinfo
rpm2cpio glx-utils-8.4.0-5.20181118git1830dcb.el8.i686.rpm | zstd -d | cpio -i --to-stdout ./usr/bin/glxinfo > glxinfo32
# Give it execute permission
chmod +x glxinfo32
# Run glxinfo within Flatpak'd Steam, and make sure that the does not say 'llvmpipe'
flatpak run --command=$PWD/glxinfo32 --filesystem=$PWD com.valvesoftware.Steam -B | grep 'OpenGL renderer string:'
```

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,9 +1,10 @@
 # Override the arch with `make ARCH=i386`
-ARCH   ?= $(shell flatpak --default-arch)
-REPO   ?= repo
+ARCH    ?= $(shell flatpak --default-arch)
+REPO    ?= repo
+FB_ARGS ?= --user --sandbox --delete-build-dirs --install-deps-from=flathub
```
Comment (TingPing) [inline Makefile:4]: Some of these args are duplicated in the call to `flatpak-builder`.

Comment (guihkx) [reply Makefile:4]: Good catch. I fixed it locally but I'll wait on more feedback before I force-push.

**Comment by TingPing** [comment]
Thanks for finishing up and testing this!

My main concerns are actually about Flathub infrastructure.

barthalion is going to look into this at some point and will ensure that this transition can go smoothly.

**Comment by guihkx** [comment]
Sync'd with `master`.

**Comment by TingPing** [comment]
Sorry about the delay but barts been busy so it might be a while.

**Comment by guihkx** [comment]
No worries at all, I completely understand.

**Comment by guihkx** [comment]
I'm thinking of opening individual pull requests for some of the commits here, because some of them are fixing actual issues which are unrelated to the move off of `extra-data`... 🤔 

@TingPing Would that be okay with you?

**Comment by TingPing** [comment]
@guihkx Sure go ahead.

@barthalion Is this still on your radar?

**Comment by Hyphaed** [comment]
anyone could help me with that? 

```
(base) ferran@z590i:~/Documents/PersonalPackages/org.freedesktop.Platform.GL.nvidia$ flatpak build-bundle repo/ nvidia-driver-560.35.03-32bit.flatpak org.freedesktop.Platform.GL32.nvidia-560-35-03
error: Refspec 'app/org.freedesktop.Platform.GL32.nvidia-560-35-03/x86_64/master' not found

```

also if I use make I do not get the 32bit version

 32bit not here :: gonna ask in the other thread

```
Remotes found with refs similar to ‘nvidia-560’:

   1) ‘flathub’ (system)
   2) ‘flathub’ (user)
   3) ‘nvidia-local’ (user)

Which do you want to use (0 to abort)? [0-3]: 3
Found ref ‘runtime/org.freedesktop.Platform.GL.nvidia-560-35-03/x86_64/1.4’ in remote ‘nvidia-local’ (user).
Use this ref? [Y/n]: 
```

**Comment by guihkx** [comment]
@Hyphaed Did you make sure to check out my branch before running the instructions? The instructions won't work if you're on the `master` branch.

**Comment by picsel2** [comment]
Let's see if I can trigger a build...

bot, build

**Comment by orowith2os** [comment]
Don't expect `bot build` to do anything, per the Flathub Matrix, from @bbhtt: 

> No, that repo is special, test builds are not supposed to work on it. Official builds from merges use a different entrypoint

**Comment by guihkx** [comment]
Rebased against current `master` to fix some conflicts.

**Comment by guihkx** [comment]
>It would make sense to build all versions at first, and the only the latest (one or few).

Right, so I think I understand the first part, but the second part I'm not really sure how you want to be achieved.

I think we could hack `build.sh` to only build drivers past specific version(s), but that might not be that obvious because NVIDIA releases many versions for multiple branches.

Perhaps another way to achieve this, would be by adding some logic to the GitHub Workflow you've introduced to use the Flathub API to determine if we should build a specific driver version or not.

In other words, you could check if ref `runtime/org.freedesktop.Platform.GL.nvidia-$VERSION/x86_64/1.4` already exists on Flathub and skip the build.

But even that solution might require manual intervention in the future, i.e. when we inevitably need to fix driver extractor (`nvidia-apply-extra.c`) because we forgot to include a library or important file.

I'm not saying it can't be done, but those are potential problems I think you guys need to be aware of before we merge this.

**Comment by guihkx** [comment]
Gotcha. So I'll mark this as ready now and you can merge at your discretion.

**Comment by guihkx** [comment]
Yes, we currently require 104 GB just for downloads, but that will only increase with time, of course.

Anyway, I have this unfinished, [testing branch](https://github.com/guihkx/org.freedesktop.Platform.GL.nvidia/tree/driver-testing) on my personal fork that I've been working on to run tests on the finished package (essentially, the idea is to ensure important files wouldn't be missing when we added a new driver version).

In the workflow, I did set up a build matrix to run on the newest version of each major release ([example run here](https://github.com/guihkx/org.freedesktop.Platform.GL.nvidia/actions/runs/13442054347)).

I think the logic in that workflow can be adapted to build *all* versions in each major release without much effort...

**Comment by guihkx** [comment]
Sure, that should work too...
